lib/Parser/Parser.cpp

```diff
@@ -1122,6 +1130,73 @@ bool Parser::parseLine(std::string &ErrStr) {
                   }
                   if (!consumeToken(ErrStr))
                     return false;
+                } else if (CurTok.str() == "range") {
+                  if (!consumeToken(ErrStr))
+                    return false;
+                  if (CurTok.K != Token::Eq) {
+                    ErrStr = makeErrStr(TP, "expected '=' for range as 'range='");
+                    return false;
+                  }
+
+                  if (!consumeToken(ErrStr))
+                    return false;
+                  if (CurTok.K != Token::OpenBracket) {
+                    ErrStr = makeErrStr(TP, "expected '[' to specify lower bound of range");
+                    return false;
+                  }
+
+                  if (!consumeToken(ErrStr))
+                    return false;
+                  if (CurTok.K != Token::UntypedInt) {
+                    ErrStr = makeErrStr(TP, "expected lower bound of range");
+                    return false;
+                  }
+                  Lower = CurTok.Val;
+
+                  if ((Lower.isNegative() && !Lower.isSignedIntN(InstWidth)) ||
+                      (!Lower.isNegative() && !(Lower.isSignedIntN(InstWidth) || Lower.isIntN(InstWidth)))) {
+                    ErrStr = makeErrStr(TP, "Lower bound is out of range");
+                    return false;
+                  }
+                  if (Lower.getBitWidth() != InstWidth)
+                    Lower = Lower.sextOrTrunc(InstWidth);
+
+                  if (!consumeToken(ErrStr))
+                    return false;
+                  if (CurTok.K != Token::Comma) {
+                    ErrStr = makeErrStr(TP, "expected ',' after lower bound of range");
+                    return false;
+                  }
+
+                  if (!consumeToken(ErrStr))
+                    return false;
+                  if (CurTok.K != Token::UntypedInt) {
+                    ErrStr = makeErrStr(TP, "expected upper bound of range");
+                    return false;
+                  }
+                  Upper = CurTok.Val;
+
+                  if ((Upper.isNegative() && !Upper.isSignedIntN(InstWidth)) ||
+                      (!Upper.isNegative() && !(Upper.isSignedIntN(InstWidth) || Upper.isIntN(InstWidth)))) {
```

Comment:
this test can be simplified by factoring out the !isSignedIntN() check -- it has to be done regardless of whether the value is negative

Reply:
Fixed. Thanks!